drivers/ethernet/eth_xilinx_axienet.c

```diff
@@ -0,0 +1,613 @@
+
+	/* inform the L2 driver about link event */
+	if (state->is_up) {
+		net_eth_carrier_on(data->interface);
+	} else {
+		net_eth_carrier_off(data->interface);
+	}
+}
+
+static void xilinx_axienet_iface_init(struct net_if *iface)
+{
+	struct xilinx_axienet_data *data = net_if_get_device(iface)->data;
+	const struct xilinx_axienet_config *config = net_if_get_device(iface)->config;
+	int err;
+
+	data->interface = iface;
+
+	ethernet_init(iface);
+
+	net_if_set_link_addr(iface, data->mac_addr, sizeof(data->mac_addr), NET_LINK_ETHERNET);
+
+	/* carrier is initially off */
+	net_eth_carrier_off(iface);
+
+	err = phy_link_callback_set(config->phy, phy_link_state_changed, data);
+
+	if (err) {
+		LOG_ERR("Could not set PHY link state changed handler : %d",
+			config->phy ? err : -1);
+	}
+
+	LOG_INF("Interface initialized!");
+}
+
+static int xilinx_axienet_send(const struct device *dev, struct net_pkt *pkt)
+{
+	struct xilinx_axienet_data *data = dev->data;
+	const struct xilinx_axienet_config *config = dev->config;
+	size_t pkt_len = net_pkt_get_len(pkt);
+	size_t current_descriptor = data->tx_populated_buffer_index;
+
+	if (net_pkt_read(pkt, data->tx_buffer[current_descriptor].buffer, pkt_len)) {
+		LOG_ERR("Failed to read packet into TX buffer!");
+		return -EIO;
+	}
+	return setup_dma_tx_transfer(dev, config, data, pkt_len);
+}
+
+static int xilinx_axienet_probe(const struct device *dev)
+{
+	const struct xilinx_axienet_config *config = dev->config;
+	struct xilinx_axienet_data *data = dev->data;
+	uint32_t status;
+	int err;
+
+	if (!config->dma) {
+		LOG_ERR("DMA handle is not provided in device tree!");
+		return -ENODEV;
+	}
+
+	status = xilinx_axienet_read_register(
+		config, XILINX_AXIENET_RECEIVER_CONFIGURATION_WORD_1_REG_OFFSET);
+	status = status & ~XILINX_AXIENET_RECEIVER_CONFIGURATION_WORD_1_REG_RX_EN_MASK;
+	xilinx_axienet_write_register(
+		config, XILINX_AXIENET_RECEIVER_CONFIGURATION_WORD_1_REG_OFFSET, status);
+
+	/* RX disabled - it is safe to modify settings */
+
+	/* clear any RX rejected interrupts from when the core was not configured */
+	xilinx_axienet_write_register(config, XILINX_AXIENET_INTERRUPT_STATUS_OFFSET,
+				      XILINX_AXIENET_INTERRUPT_STATUS_RXREJ_MASK |
+					      XILINX_AXIENET_INTERRUPT_STATUS_RXFIFOOVR_MASK);
+
+	xilinx_axienet_write_register(config, XILINX_AXIENET_INTERRUPT_ENABLE_OFFSET,
+				      config->have_irq
+					      ? XILINX_AXIENET_INTERRUPT_ENABLE_RXREJ_MASK |
+							XILINX_AXIENET_INTERRUPT_ENABLE_OVR_MASK
+					      : 0);
+
+	xilinx_axienet_write_register(config,
+				      XILINX_AXIENET_RECEIVER_CONFIGURATION_FLOW_CONTROL_OFFSET,
+				      XILINX_AXIENET_RECEIVER_CONFIGURATION_FLOW_CONTROL_EN_MASK);
+
+	err = phy_configure_link(config->phy, LINK_HALF_10BASE_T | LINK_FULL_10BASE_T |
+						      LINK_HALF_100BASE_T | LINK_FULL_100BASE_T |
+						      LINK_FULL_1000BASE_T);
+	if (err) {
+		LOG_ERR("Could not configure PHY: %d", -err);
+		return err;
+	}
```

Comment:
No support for ``LINK_HALF_1000BASE_T`` ?

Some Phys don't have this function implemented, maybe don't fail it here.

Reply:
OK, I downgraded the error to a warning.

I re-read the spec sheet - as it turns out, the AXI ethernet subsystem *does not support half duplex at all*. So I will remove it everywhere in the driver.